[sqlite] Use extended result codes

yol opened the pull request:
Extended result codes can give more information about what kind of error
occurred during an SQLite operation, which is desirable for improved
logging and debugging.

Should help with investigation of #15230

@MartijnKaijser not sure what to do with this, it's not a must for v18 but I don't know whether we can get more details on the Xbox issue without this being in the released store version.
Potential risks are that SQLite may now return different result codes in general from what was expected before. I checked the whole file and didn't find a problem, but something might have slipped through.

DaveTBlake commented on xbmc/dbwrappers/sqlitedataset.cpp:
```diff
@@ -129,56 +238,17 @@ int SqliteDatabase::status(void) {
-  error += qry;
-  error += "\n";
+int SqliteDatabase::setErr(int err_code, const char * qry) {
+  std::stringstream ss;
```
Small code style issue - `{` on new line here and elsewhere

pkerling replied:
The whole file is non-conforming so I don't think it makes much sense to change that on this line

DaveTBlake replied:
Fair enough

DaveTBlake commented:
From inspection it seems reasonable to get this into RC5, but sorry I'm not able to test it at the moment (my internet is flakey)

pkerling committed:
[sqlite] Use extended result codes

Extended result codes can give more information about what kind of error
occurred during an SQLite operation, which is desirable for improved
logging and debugging.

pkerling committed:
[sqlite] Include sqlite3_errmsg() result in error messages

Improves debugging by including information about the actual SQL error
in log messages (mostly relevant for sqlite3_prepare_v2 and
sqlite3_finalize, but doesn't hurt to just print it always)

pkerling committed:
[sqlite] Use error message returned by sqlite3_exec()

sqlite3_exec optionally gives us an error message when execution of an
SQL statement fails, but we do not use it currently.
To improve debugging, include the message in DbErrors exception message.

Also fixes a memory leak in the error path of make_query (missing
sqlite3_free) and a format string mistake (DbErrors constructor uses
printf formatting)

pkerling committed:
[sqlite] Fix DbErrors format strings

Fixes format string mistake (DbErrors constructor uses printf
formatting)

pkerling commented:
updated to just include sqlite3_errmsg() always

MartijnKaijser commented:
We could do it in the beta channel for UWP however that doesn't really have a big enough audience i think. Lets put in it first point release as that gives more time for nightly testing

pkerling commented:
> We could do it in the beta channel for UWP however that doesn't really have a big enough audience i think

Well we have some users we could ask specifically. Can they keep their settings/DB when switching to the beta channel?

pkerling commented:
Cool - can you make a build with this commit included in some channel?

pkerling commented:
@DaveTBlake Ready for merge if you approve

DaveTBlake commented:
jenkins build this please

MartijnKaijser commented:
Jenkins build this please

Rechi commented on xbmc/dbwrappers/sqlitedataset.cpp:
```diff
@@ -21,6 +23,112 @@
+  X(SQLITE_WARNING),
+  X(SQLITE_ROW),
+  X(SQLITE_DONE),
+  X(SQLITE_ERROR_MISSING_COLLSEQ),
```
> error C2065: 'SQLITE_ERROR_MISSING_COLLSEQ': undeclared identifier

some more errors about undeclared identifier on windows

pkerling replied:
What's the windows sqlite version?

Rechi replied:
lowest one is 3.10.2

MartijnKaijser replied:
Should they be updated?

pkerling replied:
+1 for update

pkerling replied:
3.10.2 is 3 years old

DaveTBlake replied:
Update for v19 sure, but to do so in point release has some risks

Rechi replied:
Ubuntu Xenial (16.04LTS) has 3.11.0, so same errors will happen there.

pkerling replied:
ok

DaveTBlake commented:
I'm not sure changing  windows sqlite version in a point release is a good idea even if 3.10.2 is 3 years old. It will certainly need some serious testing to be sure we have not been relying on some accidental behaviour.

MartijnKaijser commented on xbmc/dbwrappers/sqlitedataset.cpp:
```diff
@@ -25,6 +25,7 @@
 
 namespace {
 #define X(VAL) std::make_pair(VAL, #VAL)
+//!@todo Remove ifdefs when sqlite version requirement has been bumped
```
Perhaps add which version is expected

pkerling replied:
That means I have to crawl through all of them but yeah I guess you are right :o)

MartijnKaijser replied:
Well at least some rough guestimate. Compiler errors will show again if still not the correct version

pkerling commented:
jenkins build this please

pkerling commented:
Good to merge now from my side

MartijnKaijser commented:
Jenkins build this please
